packages/core/content-manager/server/src/controllers/collection-types.ts

```diff
@@ -301,29 +349,39 @@ export default {
       return ctx.forbidden();
     }
 
-    const permissionQuery = await permissionChecker.sanitizedQuery.publish(ctx.query);
-    // @ts-expect-error populate builder needs to be called with a UID
-    const populate = await getService('populate-builder')(model)
-      .populateFromQuery(permissionQuery)
-      .populateDeep(Infinity)
-      .countRelations()
-      .build();
+    const publishedDocument = await strapi.db.transaction(async () => {
+      // Create or update document
+      if (id) {
+        await updateDocument(ctx, { populate: {} });
+      } else {
+        const document = await createDocument(ctx, { populate: {} });
+        id = document.id;
+      }
 
-    // TODO: Publish many locales
-    const { locale = 'en' } = getDocumentDimensions(body);
-    const document = await entityManager.findOne(id, model, { populate, locale });
+      const permissionQuery = await permissionChecker.sanitizedQuery.publish(ctx.query);
+      // @ts-expect-error populate builder needs to be called with a UID
+      const populate = await getService('populate-builder')(model)
+        .populateFromQuery(permissionQuery)
+        .populateDeep(Infinity)
+        .countRelations()
+        .build();
 
-    if (!document) {
-      return ctx.notFound();
-    }
+      // TODO: Publish many locales
+      const { locale = 'en' } = getDocumentDimensions(body);
+      const document = await entityManager.findOne(id, model, { populate, locale });
```

Comment:
Is there any way we could use the returned document from L355 or 357 to avoid
calling the entityManager again here?

Reply:
ah you are right, let me take a look